spec/src/test/java/impl/UserResourceTest.java

```diff
@@ -113,15 +136,84 @@ public void shouldAddCookieToHeader() {
 
     }
 
+    @Test
+    public void shouldUpdateExistingUserIfNewPictureIsProvided() {
+        // given that the user already exists in rocket fuel
+        User user = insertUser();
+
+        // and that the user has changed profile picture since last visit
+        ImmutableOpenIdToken openId = new ImmutableOpenIdToken(user.getName(), user.getEmail(), "new_picture.png");
+        when(openIdValidator.validate(any())).thenReturn(just(openId));
+
+        // when the user signs in
+        User returnedUser = userResource.generateToken(OPEN_ID_TOKEN).toBlocking().singleOrDefault(null);
+
+        // then the user returned shall have the updated picture url
+        assertThat(returnedUser.getPicture()).isEqualTo(openId.picture);
+
+        // and the user in the database shall be updated with the new picture url
+        User storedUser = userResource.getUserById(returnedUser.getId()).toBlocking().single();
+        assertThat(storedUser.getPicture()).isEqualTo(openId.picture);
+    }
+
+    @Test
+    public void shouldUpdateExistingUserIfNewNameIsProvided() {
+        // given that the user already exists in rocket fuel
+        User user = insertUser();
+
+        // and that the user has changed name since last visit
+        ImmutableOpenIdToken openId = new ImmutableOpenIdToken("Arnold", user.getEmail(), user.getPicture());
+        when(openIdValidator.validate(any())).thenReturn(just(openId));
+
+        // when the user signs in
+        User returnedUser = userResource.generateToken(OPEN_ID_TOKEN).toBlocking().singleOrDefault(null);
+
+        // then the user returned shall have the updated name
+        assertThat(returnedUser.getName()).isEqualTo(openId.name);
+
+        // and the user in the database shall be updated with the new name
+        User storedUser = userResource.getUserById(returnedUser.getId()).toBlocking().single();
+        assertThat(storedUser.getName()).isEqualTo(openId.name);
+    }
+
+    @Test
+    public void shouldThrowInternalServerErrorIfUpdateOfUserFails() {
+        // given that the user already exists in rocket fuel
+        User user = insertUser();
+        UserDao userDao = testSetup.getInjector().getInstance(UserDao.class);
+        UserDao userDaoMock = mock(UserDao.class);
+        //but the update sql yields errors
+        doAnswer((answer)-> error(new Exception("should.not.be.executed"))).when(userDaoMock).insertUser(any());
+        doAnswer(answer -> userDao.getUserByEmail((String)answer.getArguments()[0])).when(userDaoMock).getUserByEmail(anyString());
```

Comment:
Is there a reason to execute the real getUserByEmail method instead of just mocking that as well? It makes the test a lot more complex without providing any real coverage as far as I can tell

Reply:
fixed